.github/workflows/python.yml

```diff
@@ -138,6 +138,17 @@ jobs:
         continue-on-error: true
         run: archery docker push ${{ matrix.image }}
 
+      - name: Type check with mypy and pyright
+        run: |-
+            python -m pip install mypy pyright ty griffe libcst pytest hypothesis fsspec scipy-stubs pandas-stubs types-python-dateutil types-psutil types-requests griffe libcst sphinx types-cffi
+            pip install -i https://pypi.anaconda.org/scientific-python-nightly-wheels/simple pyarrow
```

Comment:
This is currently installing pyarrow from the nightlies, as part of CI we should test the pyarrow version on the PR / commit / branch.
On this job pyarrow is built via archery on a docker container, it won't be available outside the container. We might want to add a new job only for running type checking where the docker container also runs the type checking or add it to this jobs `archery docker run`, we can probably drive the check with an environment variable and run this directly on `ci/scripts/python_test.sh`.

Reply:
Indeed. I've added [/arrow/ci/scripts/python_test_type_annotations.sh](https://github.com/apache/arrow/blob/00047e4caa4cf080080c32c5e10326ee2afec471/ci/scripts/python_test_type_annotations.sh) (and [windows](https://github.com/apache/arrow/blob/00047e4caa4cf080080c32c5e10326ee2afec471/ci/scripts/python_test_type_annotations.bat) equivalents) to call from  `compose.yaml`:
https://github.com/apache/arrow/blob/00047e4caa4cf080080c32c5e10326ee2afec471/compose.yaml#L935-L945
and python.yaml:
https://github.com/apache/arrow/blob/00047e4caa4cf080080c32c5e10326ee2afec471/.github/workflows/python.yml#L242-L246
https://github.com/apache/arrow/blob/00047e4caa4cf080080c32c5e10326ee2afec471/.github/workflows/python.yml#L304-L307